lib/vmdb/config/validator.rb

```diff
@@ -71,25 +71,16 @@ def webservices(data)
       def authentication(data)
         valid, errors = true, []
 
-        keys = data.each_pair.to_a.transpose.first.to_set
-
-        if keys.include?(:mode) && !AUTH_TYPES.include?(data.mode)
-          valid = false
-          errors << [:mode, "authentication type, \"#{data.mode}\", invalid. Should be one of: #{AUTH_TYPES.join(", ")}"]
-        end
+        return valid, errors if data.mode == 'none'
 
-        if data.mode == "ldap"
-          if data.ldaphost.blank?
-            valid = false; errors << [:ldaphost, "ldaphost can't be blank"]
-          end
-        elsif data.mode == "amazon"
-          if data.amazon_key.blank?
-            valid = false; errors << [:amazon_key, "amazon key can't be blank"]
-          end
+        authenticator = Authenticator.authenticator_class(data.mode)
 
-          if data.amazon_secret.blank?
-            valid = false; errors << [:amazon_secret, "amazon secret can't be blank"]
-          end
+        if authenticator
+          errors = authenticator.validate_config(data)
+          valid = errors.empty?
+        else
+          valid = false
+          errors << [:mode, "authentication type, \"#{data.mode}\", invalid. Should be one of: #{Authenticator::Base.subclasses.map(&:authenticates_for).join(", ")}"]
```

Comment:
Now this does not include `:amazon` because it's not a known subclass for `Authenticator::Base` until it's loaded.

```ruby
[1] pry(main)> Authenticator::Base.subclasses
=> [Authenticator::Ldap, Authenticator::Database, Authenticator::Httpd]
[2] pry(main)> Authenticator::Amazon
=> Authenticator::Amazon
[3] pry(main)> Authenticator::Base.subclasses
=> [Authenticator::Ldap, Authenticator::Amazon, Authenticator::Database, Authenticator::Httpd]
```